crates/sui-core/src/consensus_handler.rs

```diff
@@ -130,6 +132,52 @@ impl ConsensusHandlerInitializer {
     }
 }
 
+mod additional_consensus_state {
+    use super::*;
+    /// AdditionalConsensusState tracks any in-memory state that is retained by ConsensusHandler
+    /// between consensus commits. Because of crash recovery, using such data is inherently risky.
+    /// In order to do this safely, we must store data from a fixed number of previous commits.
+    /// Then, at start-up, that same fixed number of already processed commits is replayed to
+    /// reconstruct the state.
+    ///
+    /// To make sure that bugs in this process appear immediately, we record the digest of this
+    /// state in ConsensusCommitPrologue, so that any deviation causes an immediate fork.
+    #[derive(Serialize, Deserialize)]
+    pub struct AdditionalConsensusState {
+        commit_rate_estimate: CommitRateObserver,
+    }
+
+    impl AdditionalConsensusState {
+        pub fn new(additional_consensus_state_window_size: u32) -> Self {
+            Self {
+                commit_rate_estimate: CommitRateObserver::new(
+                    additional_consensus_state_window_size,
+                ),
+            }
+        }
+
+        pub fn new_for_tests() -> Self {
+            Self {
+                commit_rate_estimate: CommitRateObserver::new(10),
+            }
+        }
+
+        /// Update all internal state based on the new commit
+        pub fn observe_commit(&mut self, consensus_commit: &impl ConsensusCommitAPI) {
+            self.commit_rate_estimate
```

Comment:
does this need to be guarded by the commit number so that if a second thing is added which uses more prior commits than the rate estimator, it doesn't erroneously start uisng the extra prior commits here too?

this is somewhat nitpicky, but i bring it up bc you have a placeholder for future additions in `consensus_num_requested_prior_commits_at_startup` and no corresponding placeholder here. i could see it being missed by a future editor